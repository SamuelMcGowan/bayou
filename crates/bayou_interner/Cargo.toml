[package]
name = "bayou_interner"
version = "0.1.0"
edition = "2021"

[dependencies]
ahash = "0.8.11"
bumpalo = "3.16.0"
hashbrown = "0.14.5"

[dependencies.serde]
workspace = true
optional = true
features = ["derive"]
