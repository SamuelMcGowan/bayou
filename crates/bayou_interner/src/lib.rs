//! A string interner.
//!
//! A string interner is a data structure commonly used in compilers and other contexts that need to
//! cheaply store and compare many often identical strings. "Interning" a string returns a pointer (or in
//! this implementation, an ID) that is cheap to copy and to perform string equality checks on. This is
//! achieved by deduplicating strings using an internal hash table.
//!
//! This string interner also stores all strings in a single bump-allocated buffer, courtesy of [`bumpalo`],
//! avoiding excessive allocation.
//!
//! I decided to represent interned strings with a 32-bit ID instead of a reference to avoid introducing lifetimes.
//! This does mean that accessing the underlying string requires calling a method on the interner, but this is a
//! single array lookup.
//!
//! # Example
//! ```rust
//! use bayou_interner::Interner;
//!
//! # fn main() {
//! let interner = Interner::new();
//!
//! let hello = interner.intern("hello");
//! let hello2 = interner.intern("hello");
//! let world = interner.intern("world");
//!
//! // Interned strings can be compared cheaply.
//! assert_ne!(hello, hello2);
//! assert_ne!(hello, world);
//!
//! // Getting the associated string for an interned string.
//! assert_eq!(interner.get_str(hello), Some("hello"));
//! # }
//! ```

mod arena;

use std::cell::RefCell;
use std::num::NonZeroU32;
use std::ops::Index;

use ahash::RandomState;
use arena::InternerArena;
use hashbrown::hash_table::Entry;
use hashbrown::HashTable;

/// An ID for an interned string. Cheap to copy, and to perform string equality checks on, as
/// internally it is simply a [`NonZeroU32`] ID. It can also be stored inside an [`Option`] for free
/// due to niche optimisation.
///
/// In order to get the associated string, the interned string must be looked up
/// in the interner it was created with.
///
/// Note that performing an equality check on interned strings from different
/// interners will give a nonsensical result.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Istr(NonZeroU32);

impl Istr {
    #[inline]
    fn from_index(index: usize) -> Option<Self> {
        Some(Self(NonZeroU32::new(
            u32::try_from(index).ok()?.wrapping_add(1),
        )?))
    }

    #[inline]
    fn to_index(self) -> usize {
        self.0.get() as usize - 1
    }
}

#[derive(Clone, Copy)]
struct Metadata {
    interned: Istr,
    hash: u64,
}

#[derive(Default)]
struct Lookup {
    random_state: RandomState,
    table: HashTable<Metadata>,
}

/// Storage for interned strings.
#[derive(Default)]
pub struct Interner {
    lookup: RefCell<Lookup>,
    arena: InternerArena,
}

impl Interner {
    /// Create a new interner.
    #[inline]
    pub fn new() -> Self {
        Self::default()
    }

    /// Intern a string, returning an interned string that it is cheap to copy and
    /// perform equality checks on. Strings are only stored in the interner once, no
    /// matter how many times they are interned.
    ///
    ///
    /// ```rust
    /// # use bayou_interner::Interner;
    /// #
    /// # fn main() {
    /// let interner = Interner::new();
    ///
    /// let hello = interner.intern("hello");
    /// let hello2 = interner.intern("hello");
    /// let world = interner.intern("world");
    ///
    /// assert_eq!(hello, hello2);
    /// assert_ne!(hello, world);
    /// # }
    /// ```
    ///
    /// # Panics
    /// Panics if there are no more available IDs. An interner can store up
    /// to `u32::MAX - 1` strings before panicking.
    #[inline]
    pub fn intern(&self, key: &str) -> Istr {
        self.try_intern(key).expect("Too many interned strings")
    }

    /// Like [`Interner::intern`], but non-panicking in the case that there are no
    /// more available IDs.
    pub fn try_intern(&self, key: &str) -> Option<Istr> {
        let mut lookup = self.lookup.borrow_mut();

        let hash = lookup.random_state.hash_one(key);

        let entry = lookup.table.entry(
            hash,
            |metadata| self.arena.get(metadata.interned.to_index()) == Some(key),
            |metadata| metadata.hash,
        );

        let interned = match entry {
            Entry::Occupied(entry) => entry.get().interned,
            Entry::Vacant(entry) => {
                let index = self.arena.push_str(key);
                let interned = Istr::from_index(index)?;

                entry.insert(Metadata { interned, hash });

                interned
            }
        };

        Some(interned)
    }

    /// Get an interned string if this string is interned, otherwise return `None`.
    ///
    /// ```rust
    /// # use bayou_interner::{Interner, Istr};
    /// #
    /// # fn main() {
    /// let interner = Interner::new();
    ///
    /// let hello: Istr = interner.intern("hello");
    ///
    /// assert_eq!(interner.get_interned("hello"), Some(hello));
    /// assert_eq!(interner.get_interned("world"), None);
    /// # }
    /// ```
    pub fn get_interned(&self, key: &str) -> Option<Istr> {
        let lookup = self.lookup.borrow();

        let hash = lookup.random_state.hash_one(key);

        lookup
            .table
            .find(hash, |metadata| {
                self.arena.get(metadata.interned.to_index()) == Some(key)
            })
            .map(|metadata| metadata.interned)
    }

    /// Look up an interned string to get the associated string.
    ///
    /// Note that if the interned string was created by another interner
    /// this method will return an arbitrary string or `None`.
    ///
    /// If you know that an interned string was created by this interner, you can index into
    /// the interner instead. This is the same as calling this method, but panics if the
    /// interned string is not found in this interner instead of returning `None`.
    ///
    /// ```rust
    /// # use bayou_interner::Interner;
    /// #
    /// # fn main() {
    /// let interner = Interner::new();
    /// let interned = interner.intern("hello");
    ///
    /// assert_eq!(interner.get_str(interned), Some("hello"));
    /// # }
    /// ```
    #[inline]
    pub fn get_str(&self, interned: Istr) -> Option<&str> {
        self.arena.get(interned.to_index())
    }
}

impl Index<Istr> for Interner {
    type Output = str;

    #[inline]
    fn index(&self, interned: Istr) -> &Self::Output {
        self.get_str(interned).expect("String not in interner")
    }
}

#[test]
fn test_interner() {
    let interner = Interner::new();

    for n in 0..100 {
        let s = n.to_string();

        let a = interner.intern(&s);
        let b = interner.intern(&s);

        assert_eq!(a, b);
        assert_eq!(interner.get_str(a), Some(s.as_str()));
    }
}
