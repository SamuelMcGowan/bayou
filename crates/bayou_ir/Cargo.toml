[package]
name = "bayou_ir"
version = "0.1.0"
edition = "2021"

[dependencies]
bayou_utils = { path = "../bayou_utils" }
bayou_session = { path = "../bayou_session" }
bayou_interner = { path = "../bayou_interner" }

macro_rules_attribute = { workspace = true }
serde = { workspace = true }
