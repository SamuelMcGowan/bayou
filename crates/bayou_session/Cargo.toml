[package]
name = "bayou_session"
version = "0.1.0"
edition = "2021"

[dependencies]
bayou_common = { path = "../bayou_common" }
bayou_diagnostic = { path = "../bayou_diagnostic", features = ["serialize"] }

lasso = { version = "0.7.2", features = ["serialize"] }
serde = { workspace = true }
target-lexicon = "0.12.13"
thiserror = { workspace = true }
