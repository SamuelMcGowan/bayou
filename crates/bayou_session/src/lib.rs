pub mod diagnostics;
pub mod platform;
pub mod sourcemap;

use diagnostics::prelude::*;
use diagnostics::DiagnosticEmitter;
pub use lasso;
use sourcemap::SourceMap;
use target_lexicon::Triple;

pub type InternedStr = lasso::Spur;
pub type Interner = lasso::Rodeo;

#[derive(thiserror::Error, Debug)]
#[error("had errors")]
pub struct HadErrors;

/// Session shared between multiple package compilations.
pub struct Session<D: DiagnosticEmitter> {
    pub target: Triple,

    pub sources: SourceMap,
    pub interner: Interner,

    pub diagnostics: D,
}

impl<D: DiagnosticEmitter> Session<D> {
    pub fn new(target: Triple, diagnostics: D) -> Self {
        Self {
            target,

            sources: SourceMap::default(),
            interner: Interner::new(),

            diagnostics,
        }
    }

    pub fn report(
        &mut self,
        diagnostic: impl IntoDiagnostic,
        source_id: SourceId,
    ) -> Result<(), HadErrors> {
        let diagnostic = diagnostic.into_diagnostic(source_id, &self.interner);
        let kind = diagnostic.severity;

        self.diagnostics.emit_diagnostic(diagnostic, &self.sources);

        if kind < Severity::Error {
            Ok(())
        } else {
            Err(HadErrors)
        }
    }

    pub fn report_all<I>(&mut self, diagnostics: I, source_id: SourceId) -> Result<(), HadErrors>
    where
        I: IntoIterator,
        I::Item: IntoDiagnostic,
    {
        let mut had_error = false;

        for diagnostic in diagnostics {
            let diagnostic = diagnostic.into_diagnostic(source_id, &self.interner);
            had_error |= diagnostic.severity >= Severity::Error;
            self.diagnostics.emit_diagnostic(diagnostic, &self.sources);
        }

        if !had_error {
            Ok(())
        } else {
            Err(HadErrors)
        }
    }
}
