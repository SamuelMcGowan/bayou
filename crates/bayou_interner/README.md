A string interner.

A string interner is a data structure commonly used in compilers and other contexts that need to
cheaply store and compare many often identical strings. "Interning" a string returns a pointer (or in
this implementation, an ID) that is cheap to copy and to perform string equality checks on. This is
achieved by deduplicating strings using an internal hash table.

This string interner also stores all strings in a single bump-allocated buffer, courtesy of
[bumpalo](https://crates.io/crates/bumpalo), avoiding excessive allocation.

I decided to represent interned strings with a 32-bit ID instead of a reference to avoid introducing lifetimes.
This does mean that accessing the underlying string requires calling a method on the interner, but this is a
single array lookup.

# Example
```rust
use bayou_interner::Interner;

let interner = Interner::new();

let hello = interner.intern("hello");
let hello2 = interner.intern("hello");
let world = interner.intern("world");

// Interned strings can be compared cheaply.
assert_ne!(hello, hello2);
assert_ne!(hello, world);

// Getting the associated string for an interned string.
assert_eq!(interner.get_str(hello), Some("hello"));
```