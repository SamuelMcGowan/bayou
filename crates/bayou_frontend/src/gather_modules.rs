use bayou_interner::Interner;
use bayou_ir::IdentWithSource;
use bayou_session::{
    diagnostics::{prelude::*, sources::Source as _},
    module_loader::{ModuleLoader, ModulePath},
    sourcemap::{Source, SourceId},
    PackageSession, Session,
};

use crate::{
    ast,
    lexer::Lexer,
    module_tree::{DuplicateGlobalError, ModuleId, ModuleTree},
    parser::Parser,
    LexerError, ParseError,
};

pub enum GatherModulesError<S: Session> {
    ModuleLoaderError(<S::ModuleLoader as ModuleLoader>::Error),

    LexerError(LexerError, SourceId),
    ParseError(ParseError, SourceId),

    InvalidModuleName(IdentWithSource),
    DuplicateGlobal(DuplicateGlobalError),
}

impl<S: Session> IntoDiagnostic<Interner> for GatherModulesError<S> {
    fn into_diagnostic(self, interner: &Interner) -> bayou_session::diagnostics::Diagnostic {
        match self {
            Self::ModuleLoaderError(_) => {
                // TODO: require ModuleLoader::Error to be IntoDiagnostic
                todo!()
            }

            Self::LexerError(err, source_id) => err.into_diagnostic(&source_id),
            Self::ParseError(err, source_id) => err.into_diagnostic(&source_id),

            Self::InvalidModuleName(name) => Diagnostic::error()
                .with_message("invalid module name")
                .with_snippet(Snippet::primary(
                    format!(
                        "`{}` is not a valid name for a module",
                        interner.get_str(name.istr).unwrap()
                    ),
                    name.span.source_id,
                    name.span.span,
                )),

            Self::DuplicateGlobal(_err) => {
                // TODO: store source spans in duplicate global error
                todo!()
            }
        }
    }
}

pub struct ParsedModule {
    pub scope_id: ModuleId,
    pub source_id: SourceId,

    pub ast: ast::Module,
}

pub struct ModuleGatherer<'a, S: Session> {
    session: &'a mut S,
    package_session: &'a mut PackageSession<S>,

    errors: Vec<GatherModulesError<S>>,
}

impl<'a, S: Session> ModuleGatherer<'a, S> {
    pub fn new(session: &'a mut S, package_session: &'a mut PackageSession<S>) -> Self {
        Self {
            session,
            package_session,

            errors: vec![],
        }
    }

    pub fn run(mut self) -> (ModuleTree, Vec<ParsedModule>, Vec<GatherModulesError<S>>) {
        let mut global_scope_tree = ModuleTree::new();
        let mut parsed_modules = vec![];

        let mut modules_to_load = vec![global_scope_tree.root_id()];

        while let Some(scope_id) = modules_to_load.pop() {
            let module_path = &global_scope_tree.entry(scope_id).path;

            let Some((source_id, ast)) = self.parse_module(module_path) else {
                continue;
            };

            let submodule_names = ast.items.iter().filter_map(|item| match item {
                ast::Item::Submodule(name) => Some(name.with_source(source_id)),
                _ => None,
            });

            for submodule_name in submodule_names {
                // A submodule of the root module can't be called `main` because it would
                // clash with the root module. For now just check no modules are called `main`.
                // TODO: handle cyclic modules properly?
                if &self.package_session.interner[submodule_name.istr] == "main" {
                    // span is in
                    self.errors
                        .push(GatherModulesError::InvalidModuleName(submodule_name));
                    continue;
                }

                let submodule_id =
                    match global_scope_tree.insert_module(scope_id, submodule_name.istr) {
                        Ok(id) => id,
                        Err(err) => {
                            self.errors.push(GatherModulesError::DuplicateGlobal(err));
                            continue;
                        }
                    };

                modules_to_load.push(submodule_id);
            }

            parsed_modules.push(ParsedModule {
                scope_id,
                source_id,
                ast,
            });
        }

        (global_scope_tree, parsed_modules, self.errors)
    }

    fn parse_module(&mut self, module_path: &ModulePath) -> Option<(SourceId, ast::Module)> {
        let source_string = match self
            .package_session
            .module_loader
            .load_module(module_path, &self.package_session.interner)
        {
            Ok(s) => s,
            Err(err) => {
                self.errors.push(GatherModulesError::ModuleLoaderError(err));
                return None;
            }
        };

        let (source_id, source) = self.session.source_map_mut().insert_and_get(Source {
            name: module_path
                .display(&self.package_session.interner)
                .to_string(),
            source: source_string,
        });

        let (tokens, lexer_errors) =
            Lexer::new(source.source_str(), &self.package_session.interner).lex();
        self.errors.extend(
            lexer_errors
                .into_iter()
                .map(|err| GatherModulesError::LexerError(err, source_id)),
        );

        let (ast, parse_errors) = Parser::new(tokens).parse();
        self.errors.extend(
            parse_errors
                .into_iter()
                .map(|err| GatherModulesError::ParseError(err, source_id)),
        );

        Some((source_id, ast))
    }
}
