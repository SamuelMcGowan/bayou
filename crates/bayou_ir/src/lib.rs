//! Types for representing the program throughout the compilation pipeline.

#[macro_use]
extern crate macro_rules_attribute;

pub mod ir;
pub mod symbols;

use bayou_diagnostic::span::Span;
use bayou_session::InternedStr;

derive_alias! {
    #[derive(NodeTraits!)] = #[derive(Debug, Clone, PartialEq, Eq, serde::Serialize)];
    #[derive(NodeCopyTraits!)] = #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, serde::Serialize)];
}

#[derive(NodeCopyTraits!)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,

    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    // Eq,
    // NotEq,

    // Gt,
    // Lt,
    // GtEq,
    // LtEq,
}

#[derive(NodeCopyTraits!)]
pub enum UnOp {
    Negate,
    BitwiseInvert,
}

#[derive(NodeCopyTraits!)]
pub enum Type {
    I64,
    Bool,
    Void,
    Never,
}

#[derive(NodeCopyTraits!)]
pub struct Spanned<T> {
    pub node: T,
    pub span: Span,
}

impl<T> Spanned<T> {
    pub fn new(node: T, span: Span) -> Self {
        Self { node, span }
    }
}

impl<T, E> Spanned<Result<T, E>> {
    pub fn transpose(self) -> Result<Spanned<T>, E> {
        self.node.map(|node| Spanned {
            node,
            span: self.span,
        })
    }
}
