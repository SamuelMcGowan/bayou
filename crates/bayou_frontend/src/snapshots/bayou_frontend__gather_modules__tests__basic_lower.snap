---
source: crates/bayou_frontend/src/gather_modules.rs
info: "func main() -> i64 { return 0; }"
---
- items:
    - FuncDecl:
        id: 0
        block:
          statements:
            - Return:
                kind:
                  Constant:
                    I64: 0
                span:
                  span:
                    start: 28
                    end: 29
                  source_id: 0
                ty: ~
          final_expr:
            kind:
              Constant: Void
            span:
              span:
                start: 31
                end: 32
              source_id: 0
            ty: ~
          span:
            span:
              start: 19
              end: 32
            source_id: 0
  main_func: 0
- locals: {}
  funcs:
    0:
      ident:
        istr: 1
        span:
          span:
            start: 5
            end: 9
          source_id: 0
      ret_ty: I64
      ret_ty_span:
        span:
          start: 15
          end: 18
        source_id: 0
- []
